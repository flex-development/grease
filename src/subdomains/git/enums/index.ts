/**
 * @file Entry Point - Enums
 * @module grease/git/enums
 */

export { default as RepositoryProvider } from './repository-provider'
