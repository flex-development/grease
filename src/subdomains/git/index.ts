/**
 * @file Entry Point - git
 * @module grease/git
 */

export * from './enums'
export { default as GitModule } from './git.module'
export type * from './interfaces'
export * from './models'
export * from './options'
export * from './providers'
export type * from './types'
