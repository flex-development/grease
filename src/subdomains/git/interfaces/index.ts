/**
 * @file Entry Point - Interfaces
 * @module grease/git/interfaces
 */

export type { default as ICommit } from './commit.interface'
