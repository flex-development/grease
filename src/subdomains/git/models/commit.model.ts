/**
 * @file Models - Commit
 * @module grease/git/models/Commit
 */

import type { RepositoryProvider } from '#src/git/enums'
import type { ICommit } from '#src/git/interfaces'
import type { CommitGrammarOptions } from '#src/git/options'
import type {
  Author,
  BreakingChange,
  CommitFields,
  Reference,
  Trailer
} from '#src/git/types'
import {
  at,
  cast,
  construct,
  defaults,
  get,
  ifelse,
  objectify,
  pick,
  select,
  split,
  trim,
  trimEnd,
  unique,
  type Nullable,
  type Simplify
} from '@flex-development/tutils'
import CommitGrammar from './commit-grammar.model'
import Repository from './repository.model'

/**
 * Parsed commit model.
 *
 * @class
 * @implements {ICommit}
 */
class Commit implements ICommit {
  /**
   * Raw commit chunk.
   *
   * @protected
   * @readonly
   * @instance
   * @member {string} chunk
   */
  protected readonly chunk: string

  /**
   * Raw commit fields object.
   *
   * @see {@linkcode CommitFields}
   *
   * @protected
   * @readonly
   * @instance
   * @member {CommitFields} fields
   */
  protected readonly fields: CommitFields

  /**
   * Commit grammar instance.
   *
   * @see {@linkcode CommitGrammar}
   *
   * @public
   * @readonly
   * @instance
   * @member {CommitGrammar} grammar
   */
  public readonly grammar: CommitGrammar

  /**
   * Repository provider name.
   *
   * @public
   * @readonly
   * @instance
   * @member {RepositoryProvider} provider
   */
  public readonly provider: RepositoryProvider

  /**
   * Parent repository.
   *
   * @public
   * @readonly
   * @instance
   * @member {Repository} repository
   */
  public readonly repository: Repository

  /**
   * Create a new parsed commit.
   *
   * @param {Buffer | Commit | string} chunk - Raw commit chunk or parsed commit
   * @param {Repository?} [repository=new Repository()] - Parent repository
   * @param {Partial<CommitGrammarOptions>?} [grammar] - Commit grammar options
   */
  constructor(
    chunk: Buffer | Commit | string,
    repository: Repository = new Repository(),
    grammar?: Partial<CommitGrammarOptions>
  ) {
    this.chunk = (chunk = chunk.toString()).padEnd(chunk.length + 1, '\n')
    if (!trim(this.chunk)) throw new Error('empty commit chunk')

    // set repository and provider
    this.repository = get(chunk, 'repository', repository)
    this.provider = this.repository.provider

    // set commit grammar
    this.grammar = get(chunk, 'grammar', new CommitGrammar({
      ...grammar,
      provider: this.provider
    }))

    // init commit fields object
    this.fields = construct(cast<CommitFields>(objectify(
      [...this.chunk.matchAll(this.grammar.field)],
      match => get(match, 'groups.field', ''),
      match => trim(get(match, 'groups.value', ''))
    )))

    // add header data to commit fields object
    this.fields = defaults(
      this.fields,
      pick(get(this.grammar.header.exec(this.fields.header), 'groups')!, [
        'breaking',
        'pr',
        'scope',
        'subject',
        'type'
      ]),
      {
        breaking: '',
        pr: 'null',
        scope: '',
        subject: '',
        type: ''
      }
    )

    // remove extra chunk padding
    this.chunk = this.chunk.slice(0, -1)
  }

  /**
   * Commit author details.
   *
   * @see {@linkcode Author}
   *
   * @public
   *
   * @return {Author} Commit author object
   */
  public get author(): Author {
    return this.fields.author
  }

  /**
   * Commit body text.
   *
   * @public
   *
   * @return {string} Commit body
   */
  public get body(): string {
    return trimEnd(this.fields.body.replace(this.fields.trailers, ''))
  }

  /**
   * Commit contains breaking changes?
   *
   * @public
   *
   * @return {boolean} Boolean indicating commits contains breaking changes
   */
  public get breaking(): boolean {
    return this.breaks.length > 0
  }

  /**
   * Breaking changes noted in {@linkcode subject} and {@linkcode trailers}.
   *
   * @see {@linkcode BreakingChange}
   *
   * @public
   *
   * @return {BreakingChange<Commit>[]} Breaking changes array
   */
  public get breaks(): BreakingChange<Commit>[] {
    return unique(select(
      [
        this.fields.breaking
          ? {
            pr: this.pr,
            scope: this.scope,
            token: 'BREAKING-CHANGE',
            type: this.type,
            value: this.subject
          }
          : <Trailer>{ token: this.fields.breaking },
        ...this.trailers
      ],
      trailer => /^BREAKING[ -]CHANGE/.test(trailer.token),
      trailer => ({
        commit: this,
        mentions: this.mentions,
        pr: get(trailer, 'pr', null),
        scope: get(trailer, 'scope', null),
        subject: trailer.value,
        type: this.type
      })
    ))
  }

  /**
   * Commit date in strict ISO 8601 format (`%cI`).
   *
   * @see https://git-scm.com/docs/pretty-formats/2.21.0
   *
   * @public
   *
   * @return {string} Commit date
   */
  public get date(): string {
    return this.fields.date
  }

  /**
   * Abbreviated commit {@linkcode sha}.
   *
   * @public
   *
   * @return {string} Commit hash
   */
  public get hash(): string {
    return this.fields.hash
  }

  /**
   * Commit {@linkcode type}, {@linkcode scope}, breaking change indicator, and
   * {@linkcode subject}.
   *
   * @public
   *
   * @return {string} Commit header text
   */
  public get header(): string {
    return this.fields.header
  }

  /**
   * Users mentioned in commit.
   *
   * @public
   *
   * @return {string[]} Mentions array
   */
  public get mentions(): string[] {
    return select(
      [...this.chunk.matchAll(this.grammar.mention)],
      null,
      match => get(match, 'groups.user')!
    )
  }

  /**
   * Pull request number.
   *
   * @public
   *
   * @return {Nullable<number>} Pull request number or `null`
   */
  public get pr(): Nullable<number> {
    return cast(JSON.parse(this.fields.pr))
  }

  /**
   * Issue and pull request references.
   *
   * @see {@linkcode Reference}
   *
   * @public
   *
   * @return {Reference[]} Reference array
   */
  public get references(): Reference[] {
    return select(
      [...this.chunk.matchAll(this.grammar.reference)],
      null,
      m => ({
        action: get(m, 'groups.action', null),
        number: JSON.parse(get(m, 'groups.number', 'null')),
        owner: get(m, 'groups.owner', null),
        ref: get(m, 'groups.ref', ''),
        repo: get(m, 'groups.repo', null)
      })
    )
  }

  /**
   * Commit scope.
   *
   * @public
   *
   * @return {Nullable<string>} Commit scope or `null`
   */
  public get scope(): Nullable<string> {
    return ifelse(this.fields.scope, this.fields.scope, null)
  }

  /**
   * Commit sha.
   *
   * @public
   *
   * @return {string} Commit sha
   */
  public get sha(): string {
    return this.fields.sha
  }

  /**
   * Commit subject.
   *
   * @public
   *
   * @return {string} Commit subject text
   */
  public get subject(): string {
    return this.fields.subject
  }

  /**
   * Tags associated with commit.
   *
   * @public
   *
   * @return {string[]} Tags array
   */
  public get tags(): string[] {
    return select(
      select(
        split(this.fields.tags, ','),
        tag => !!tag.length,
        tag => trim(tag.replace(/^tag: */, ''))
      ),
      tag => this.grammar.tag.test(tag)
    )
  }

  /**
   * Commit message trailers.
   *
   * @public
   *
   * @return {Trailer[]} Trailers array
   */
  public get trailers(): Trailer[] {
    return select(
      [...this.fields.trailers.matchAll(this.grammar.trailer)],
      null,
      match => pick(match.groups!, ['token', 'value'])
    )
  }

  /**
   * Commit type.
   *
   * @public
   *
   * @return {string} Commit type
   */
  public get type(): string {
    return this.fields.type
  }

  /**
   * Tagged commit version.
   *
   * @public
   *
   * @return {Nullable<string>} First tag in {@linkcode tags} or `null`
   */
  public get version(): Nullable<string> {
    return at(this.tags, 0, null)
  }

  /**
   * Get a JSON-serializable commit object.
   *
   * @public
   *
   * @return {Simplify<ICommit>} JSON-serializable commit
   */
  public toJSON(): Simplify<ICommit> {
    return {
      author: this.author,
      body: this.body,
      breaking: this.breaking,
      breaks: this.breaks,
      date: this.date,
      hash: this.hash,
      header: this.header,
      mentions: this.mentions,
      pr: this.pr,
      references: this.references,
      scope: this.scope,
      sha: this.sha,
      subject: this.subject,
      tags: this.tags,
      trailers: this.trailers,
      type: this.type,
      version: this.version
    }
  }

  /**
   * Get `this` raw commit.
   *
   * @public
   *
   * @return {string} Raw commit string
   */
  public toString(): string {
    return this.chunk
  }
}

export default Commit
