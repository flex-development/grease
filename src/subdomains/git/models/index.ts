/**
 * @file Entry Point - Models
 * @module grease/git/models
 */

export { default as CommitGrammar } from './commit-grammar.model'
export { default as Commit } from './commit.model'
export { default as Repository } from './repository.model'
