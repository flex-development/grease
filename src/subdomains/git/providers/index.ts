/**
 * @file Entry Point - Providers
 * @module grease/git/providers
 */

export { default as GitService } from './git.service'
