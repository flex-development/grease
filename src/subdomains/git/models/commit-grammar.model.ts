/**
 * @file Models - CommitGrammar
 * @module grease/git/models/CommitGrammar
 */

import { CommitGrammarOptions } from '#src/git/options'
import {
  join,
  regexp,
  select,
  template,
  unique,
  type Partial
} from '@flex-development/tutils'

/**
 * Commit parser grammar.
 *
 * @class
 */
class CommitGrammar {
  /**
   * Grammar options.
   *
   * @see {@linkcode CommitGrammarOptions}
   *
   * @protected
   * @readonly
   * @instance
   * @member {CommitGrammarOptions} options
   */
  protected readonly options: CommitGrammarOptions

  /**
   * Create a new commit grammar instance.
   *
   * @see {@linkcode CommitGrammarOptions}
   *
   * @param {Partial<CommitGrammarOptions>?} [options] - Grammar options
   */
  constructor(options?: Partial<CommitGrammarOptions>) {
    this.options = new CommitGrammarOptions(options)
  }

  /**
   * Regular expression matching fields in a raw commit.
   *
   * @see https://regex101.com/r/QnFMsV
   *
   * @public
   *
   * @return {RegExp} Raw commit field regex
   */
  public get field(): RegExp {
    return /^-(?<field>[^\s$]+?)-\n(?<value>.*?(?=\n*^-\S+?-)|(?:.*$))/gms
  }

  /**
   * Regular expression matching a commit header.
   *
   * @see https://regex101.com/r/WAJeLp
   * @see https://conventionalcommits.org
   *
   * @public
   *
   * @return {RegExp} Commit header regex
   */
  public get header(): RegExp {
    return /^(?<type>[a-z]+)(?:\((?<scope>[a-z-]+)\))?(?<breaking>!)?: +(?<subject>(?:.+ \([!#](?<pr>\d+)\))|.+)/i
  }

  /**
   * Regular expression matching mentions (e.g. `@unicornware`) in a raw commit.
   *
   * @see https://regex101.com/r/upbRpj
   *
   * @public
   *
   * @return {RegExp} Mention regex
   */
  public get mention(): RegExp {
    return /\B(?<mention>@(?<user>[\w-]{1,38}(?=[^\w-]\B|\s)))/g
  }

  /**
   * Get a regular expression matching references in a raw commit.
   *
   * @see https://regex101.com/r/Thsp1M
   *
   * @public
   *
   * @return {RegExp} Reference regex
   */
  public get reference(): RegExp {
    /**
     * Pattern matching issue references in commit bodies.
     *
     * @const {string} pattern
     */
    const pattern: string =
      '(?:(?:(?<action>{actions}) +)|(?<repository>(?<owner>[\\da-z](?:-(?=[\\da-z])|[\\da-z]){0,38}(?<=[\\da-z]))\\/(?<repo>\\S+)))?(?<ref>(?<prefix>{prefixes})(?<number>\\d+))'

    return new RegExp(
      template(pattern, {
        actions: join(this.options.actions, '|'),
        prefixes: join(select(unique([
          ...this.options.issues,
          ...this.options.pr
        ]), null, regexp), '|')
      }),
      'gi'
    )
  }

  /**
   * Regular expression matching a semantic release tag.
   *
   * @see https://regex101.com/r/WyQyHZ
   * @see https://semver.org
   *
   * @public
   *
   * @return {RegExp} Semantic release tag regex
   */
  public get tag(): RegExp {
    /**
     * Pattern matching semantic release tags.
     *
     * @const {string} pattern
     */
    const pattern: string =
      '^(?<prefix>{prefix})(?<version>(?<major>0|[1-9]\\d*)\\.(?<minor>0|[1-9]\\d*)\\.(?<patch>0|[1-9]\\d*)(?:-(?<prerelease>(?:0|[1-9]\\d*|\\d*[A-Za-z-][\\dA-Za-z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[A-Za-z-][\\dA-Za-z-]*))*))?(?:\\+(?<metadata>[\\dA-Za-z-]+(?:\\.[\\dA-Za-z-]+)*))?)$'

    return new RegExp(template(pattern, { prefix: this.options.tagprefix }))
  }

  /**
   * Regular expression matching trailers in a raw commit.
   *
   * @see https://git-scm.com/docs/git-interpret-trailers
   * @see https://regex101.com/r/I56Kgg
   *
   * @return {RegExp} Git trailer regex
   */
  public get trailer(): RegExp {
    return /(?<=^|\n)(?<token>[\w -]+?(?=:)): (?<value>[\S\s]+?(?:(?=\n[\w -]+?(?=:))|(?=\n*$)))/g
  }
}

export default CommitGrammar
