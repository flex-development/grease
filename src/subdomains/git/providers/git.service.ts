/**
 * @file Providers - GitService
 * @module grease/git/providers/GitService
 */

import type { ICommit } from '#src/git/interfaces'
import { Commit, Repository } from '#src/git/models'
import {
  GitCommitOptions,
  GitOptions,
  GitTagOptions
} from '#src/git/options'
import { LoggerService, ValidationService } from '#src/providers'
import {
  at,
  ifelse,
  isNull,
  join,
  pick,
  regexp,
  select,
  sift,
  sort,
  split,
  template,
  timeunix,
  trim,
  tryit,
  type Fn,
  type Nullable
} from '@flex-development/tutils'
import { Injectable } from '@nestjs/common'
import { exec, type ExecException, type ExecOptions } from 'node:child_process'
import util from 'node:util'
import semver from 'semver'

/**
 * Git operations provider.
 *
 * @see https://git-scm.com/docs
 *
 * @class
 */
@Injectable()
class GitService {
  /**
   * String used to separate commit logs.
   *
   * @public
   * @static
   * @readonly
   * @member {string} RAW_COMMIT_DELIMITER
   */
  public static readonly RAW_COMMIT_DELIMITER: string = join([
    '--',
    '$',
    '--'
  ], '')

  /**
   * Commit log format.
   *
   * @see https://git-scm.com/docs/git-log#_pretty_formats
   *
   * @public
   * @static
   * @readonly
   * @member {string} RAW_COMMIT_FORMAT
   */
  public static readonly RAW_COMMIT_FORMAT: string = join([
    '',
    'author.email',
    '%n%ae%n',
    'author.name',
    '%n%an%n',
    'body',
    '%n%b%n',
    'date',
    '%n%cI%n',
    'hash',
    '%n%h%n',
    'header',
    '%n%s%n',
    'sha',
    '%n%H%n',
    'tags',
    '%n%D%n',
    'trailers',
    '%n%(trailers)%n'
  ], '-')

  /**
   * Create a new git operations service.
   *
   * @see {@linkcode LoggerService}
   * @see {@linkcode ValidationService}
   *
   * @param {LoggerService} logger - Logger service
   * @param {ValidationService} validator - Validation service
   */
  constructor(
    protected readonly logger: LoggerService,
    protected readonly validator: ValidationService
  ) {
    this.logger = logger.withTag('git')
  }

  /**
   * Get an array of parsed commits.
   *
   * @see {@linkcode GitCommitOptions}
   *
   * @template T - Parsed commit type
   *
   * @public
   * @async
   *
   * @param {Partial<GitCommitOptions<T>>?} [opts] - Commit options
   * @return {Promise<T[]>} Parsed commit array
   */
  public async commits<T extends Commit = Commit>(
    opts?: Partial<GitCommitOptions<T>>
  ): Promise<T[]> {
    const {
      Commit,
      cwd,
      from,
      grammar,
      to
    } = await this.validator.validate(opts = new GitCommitOptions<T>(opts))

    /**
     * Repository instance.
     *
     * @const {Repository} repository
     */
    const repository: Repository = new Repository(await this.origin(opts))

    /**
     * Raw commit logs.
     *
     * @const {string} logs
     */
    const logs: string = await this.log(sift([
      '--decorate-refs=refs/tags',
      '--decorate=short',
      template('--format=\'{format}{delimiter}\'', {
        delimiter: GitService.RAW_COMMIT_DELIMITER,
        format: GitService.RAW_COMMIT_FORMAT
      }),
      join(sift([from, to]), '..'),
      ifelse(cwd === process.cwd(), '', `-- ${cwd}`)
    ]), opts)

    return sort(select(
      split(logs, new RegExp(`${regexp(GitService.RAW_COMMIT_DELIMITER)}\n?`)),
      chunk => !!trim(chunk),
      chunk => new Commit(trim(chunk), repository, grammar)
    ), (a, b) => timeunix(b.date) - timeunix(a.date))
  }

  /**
   * Execute a command.
   *
   * Throws if the command fails.
   *
   * @public
   * @async
   *
   * @param {ReadonlyArray<string>} args - Command arguments
   * @param {Partial<GitOptions>?} [opts={}] - Git exec options
   * @return {Promise<string>} Command output
   * @throws {Error} If command fails
   */
  public async exec(
    args: readonly string[],
    opts: Partial<GitOptions> = {}
  ): Promise<string> {
    /**
     * Command to execute.
     *
     * @const {string} command
     */
    const command: string = join(['git', ...args], ' ')

    // debug command
    this.logger.sync(opts).debug(command)

    // execute command
    const [e, result] = await tryit<
      Fn<[string, ExecOptions], Promise<{ stdout: string }>>,
      ExecException & { stderr: string }
    >(util.promisify(exec))(command, {
      ...opts,
      maxBuffer: Number.POSITIVE_INFINITY,
      shell: process.env.SHELL
    })

    // throw if command failed
    if (e) {
      e.stderr = trim(e.stderr.replace(/^fatal: */, ''))
      throw new Error(e.stderr, { cause: pick(e, ['cmd', 'code']) })
    }

    return trim(result.stdout)
  }

  /**
   * Get commit logs.
   *
   * @see https://git-scm.com/docs/git-log
   *
   * @public
   * @async
   *
   * @param {ReadonlyArray<string>?} [args=[]] - `git log` options
   * @param {Partial<GitOptions>?} [opts={}] - Git exec options
   * @return {Promise<string>} Command output
   */
  public async log(
    args: readonly string[] = [],
    opts: Partial<GitOptions> = {}
  ): Promise<string> {
    return this.exec(['log', ...args], opts)
  }

  /**
   * Get a remote origin url.
   *
   * @public
   * @async
   *
   * @param {Partial<GitOptions>?} [opts={}] - Git exec options
   * @return {Promise<string>} Remote origin url
   */
  public async origin(opts?: Partial<GitOptions>): Promise<string> {
    return this.exec(['config', '--get', 'remote.origin.url'], opts)
  }

  /**
   * Get the name of the oldest tag containing `commit`.
   *
   * @see {@linkcode GitTagOptions}
   * @see {@linkcode ICommit}
   *
   * @public
   * @async
   *
   * @param {Pick<ICommit, 'sha'>} commit - Commit to find
   * @param {Partial<GitTagOptions>} [opts] - Search options
   * @return {Promise<Nullable<string>>} Tag containing `commit` or `null`
   * @throws {Error}
   */
  public async parent(
    commit: Pick<ICommit, 'sha'>,
    opts?: Partial<GitTagOptions>
  ): Promise<Nullable<string>> {
    opts = await this.validator.validate(new GitTagOptions(opts))

    // get chronological list of tags containing commit
    const [e, tags] = await tryit(this.exec.bind(this))([
      'tag',
      '--sort creatordate',
      template('--list \'{tagprefix}*\'', opts),
      template('--contains {sha}', commit)
    ], opts)

    // throw fatal error
    /* c8 ignore next */ if (e && !/no such commit/.test(e.message)) throw e

    return ifelse(isNull(e) && tags, at(split(tags, '\n'), 0), null)
  }

  /**
   * Get an array containing all git tags in reverse chronological order.
   *
   * @see {@linkcode GitTagOptions}
   *
   * @public
   * @async
   *
   * @param {Partial<GitTagOptions>?} [opts] - Git tag options
   * @return {Promise<string[]>} Git tags array
   */
  public async tags(opts?: Partial<GitTagOptions>): Promise<string[]> {
    opts = await this.validator.validate(new GitTagOptions(opts))

    /**
     * Raw tags.
     *
     * @const {string} tags
     */
    const tags: string = await this.exec([
      'tag',
      '--sort -creatordate',
      template('--list \'{tagprefix}*\'', opts)
    ], opts)

    return select(sift(select(split(tags, '\n'), null, trim)), tag => {
      return opts!.unstable
        ? true
        : !semver.parse(tag.replace(opts!.tagprefix!, ''))!.prerelease.length
    })
  }
}

export default GitService
