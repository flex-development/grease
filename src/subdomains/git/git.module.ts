/**
 * @file GitModule
 * @module grease/git/GitModule
 */

import { LoggerService, ValidationService } from '#src/providers'
import { Global, Module } from '@nestjs/common'
import { GitService } from './providers'

/**
 * Global git operations module.
 *
 * @class
 */
@Global()
@Module({
  exports: [GitService],
  providers: [GitService, LoggerService, ValidationService]
})
class GitModule {}

export default GitModule
