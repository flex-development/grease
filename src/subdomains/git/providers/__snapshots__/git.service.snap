// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:git/providers/GitService > #tags > should return tags array with unstable tags 1`] = `
[
  "1.0.0-alpha.23",
  "1.0.0-alpha.22",
  "1.0.0-alpha.21",
  "1.0.0-alpha.20",
  "1.0.0-alpha.19",
  "1.0.0-alpha.18",
  "1.0.0-alpha.17",
  "1.0.0-alpha.16",
  "1.0.0-alpha.15",
  "1.0.0-alpha.14",
  "1.0.0-alpha.13",
  "1.0.0-alpha.12",
  "1.0.0-alpha.11",
  "1.0.0-alpha.10",
  "1.0.0-alpha.9",
  "1.0.0-alpha.8",
  "1.0.0-alpha.7",
  "1.0.0-alpha.6",
  "1.0.0-alpha.5",
  "1.0.0-alpha.4",
  "1.0.0-alpha.3",
  "1.0.0-alpha.2",
  "1.0.0-alpha.1",
]
`;

exports[`unit:git/providers/GitService > #tags > should return tags array without unstable tags 1`] = `
[
  "tutils@5.0.1",
  "tutils@5.0.0",
  "tutils@4.8.0",
  "tutils@4.7.0",
  "tutils@4.6.0",
  "tutils@4.5.0",
  "tutils@4.4.0",
  "tutils@4.3.0",
  "tutils@4.2.3",
  "tutils@4.2.2",
  "tutils@4.2.1",
  "tutils@4.2.0",
  "tutils@4.1.1",
  "tutils@4.1.0",
  "tutils@4.0.3",
  "tutils@4.0.2",
  "tutils@4.0.1",
  "tutils@4.0.0",
  "tutils@3.1.7",
  "tutils@3.1.6",
  "tutils@3.1.5",
  "tutils@3.1.4",
  "tutils@3.1.3",
  "tutils@3.1.2",
  "tutils@3.1.1",
  "tutils@3.1.0",
  "tutils@3.0.0",
  "tutils@2.0.0",
  "tutils@1.0.0",
]
`;
